lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -613,9 +613,34 @@ void TypeLoweringVisitor::processUsers(Value val, ArrayRef<Value> mapping) {
       sfo.replaceAllUsesWith(repl);
       sfo.erase();
     } else {
-      val.dump();
-      val.getDefiningOp()->getParentOfType<FModuleOp>()->dump();
-      llvm_unreachable("Unknown aggregate user");
+      // This means, we have already processed the user, and it didnot lower its
+      // inputs. This is an opaque user, which will continue to have aggregate
+      // type as input, even after LowerTypes. So, construct the vector/bundle
+      // back from the lowered elements to ensure a valid input into the opaque
+      // op. This only supports Bundle or vector of ground type elements.
+      // Recursive aggregate types are not yet supported.
+
+      // This builder ensures that the aggregate construction happens at the
+      // user location, and the LowerTypes algorithm will not touch them any
+      // more, because LowerTypes was reverse iterating on the block and the
+      // user has already been processed.
+      ImplicitLocOpBuilder b(user->getLoc(), user);
+      // Get the operand number in the op.
+      auto it = llvm::find(user->getOperands(), val);
+      // Cat all the field elements.
+      Value accumulate;
+      for (auto v : mapping) {
+        if (!v.getType().cast<FIRRTLBaseType>().isGround()) {
+          user->emitError("cannot handle an opaque user of aggregate types "
+                          "with non-ground type elements");
+          return;
+        }
```

Comment:
Probably this check is not required. We might want to preserve multi-dim vectors.

Reply:
We can use `bitcast(cooncat(bitcast(op1), ..., bitcast(op_n)))` but I'm fine with keeping current implementation as is.